### PR TITLE
Bug #14015-6.3.x: fixing integrity problems into administration caches.

### PR DESCRIPTION
**Pull request by SilverYoCha**
Refactoring SpaceInstLazyDataLoader in order to ensure providing sub space instances and component instances from administration caches when using SpaceInst.getSubSpaces() or SpaceInst.getAllComponentInst().

The space copy/paste function has been slightly reworked to avoid using SpaceInstLazyLoader as a data container for new data to be persisted as much as possible.

### CODE REVIEW

File: core-library/src/main/java/org/silverpeas/core/admin/space/SpaceInst.java [hunk 1]
```diff
@@ -351,9 +326,11 @@ public int getNumComponentInst() {
...
    */
   public void addComponentInst(ComponentInst componentInst) {
-    data.safeWrite((p, s, c) -> {
-      c.removeIf(i -> i.getId().equals(componentInst.getId()));
-      c.add(componentInst);
+    data.safeWrite(d -> {
+      final String cId = componentInst.getId();
+      final List<String> c = d.getComponentIds();
+      c.removeIf(cId::equals);
+      c.add(cId);
```
Comment (mmoqui) [inline core-library/src/main/java/org/silverpeas/core/admin/space/SpaceInst.java:333]: Why removing the component instance id if it is already in the list? Shouldn't be added only if it isn't yet in the list?

File: core-library/src/main/java/org/silverpeas/core/admin/space/SpaceInstLazyDataLoader.java [hunk 2]
```diff
@@ -76,19 +92,55 @@ public class SpaceInstLazyDataLoader implements Serializable {
...
+   */
+  void copy(final SpaceInstLazyDataLoader other) {
+    final List<String> componentIdsToCopy;
+    final List<String> subSpaceIdsToCopy;
+    final List<SpaceProfileInst> spaceProfilesToCopy;
+    synchronized (other.mutex) {
+      // loading data if not yet done
+      if (!other.loaded) {
+        other.loaded = true;
+        other.load();
```
Comment (mmoqui) [inline core-library/src/main/java/org/silverpeas/core/admin/space/SpaceInstLazyDataLoader.java:113]: the `other.loaded` attribute shouldn't be set at true by the `other.load()` method itself once the loading done?

File: core-library/src/main/java/org/silverpeas/core/admin/service/DefaultAdministration.java [hunk 3]
```diff
@@ -380,21 +380,10 @@ public String addSpaceInst(String userId, SpaceInst spaceInst) throws AdminExcep
...
-
-      // Instantiate the components
-      List<ComponentInst> alCompoInst = spaceInst.getAllComponentsInst();
-      for (ComponentInst componentInst : alCompoInst) {
-        componentInst.setDomainFatherId(spaceInst.getId());
-        addComponentInst(userId, componentInst);
-      }
-
-      SpaceInstLight space = getSpaceInstLight(spaceInst.getLocalId());
+      final SpaceInstLight space = getSpaceInstLight(spaceInst.getLocalId());
```
[the remark this answers is not in the document]
Comment (SilverYoCha) [reply core-library/src/main/java/org/silverpeas/core/admin/service/DefaultAdministration.java:383]: `cache.opAddSpace(getSpaceInstById(spaceInst.getLocalId()));` is deealing with a `SpaceInst` instance.

Whereas `final SpaceInstLight space = getSpaceInstLight(spaceInst.getLocalId());` is dealing with a `SpaceInstLight` instance.

File: core-library/src/main/java/org/silverpeas/core/admin/space/SpaceInst.java (hunk 1 above)
Comment (SilverYoCha) [reply core-library/src/main/java/org/silverpeas/core/admin/space/SpaceInst.java:333]: It is expected here that the old representation of a component instance is removed before again added.

Comment (mmoqui) [reply core-library/src/main/java/org/silverpeas/core/admin/space/SpaceInst.java:333]: But why added it again if it is already in the list? Why not keeping it in the list and add it only if it isn't in the list?

File: core-library/src/main/java/org/silverpeas/core/admin/space/SpaceInstLazyDataLoader.java (hunk 2 above)
Comment (mmoqui) [reply core-library/src/main/java/org/silverpeas/core/admin/space/SpaceInstLazyDataLoader.java:113]: Hum  ... not so convinced. But ok

File: core-library/src/main/java/org/silverpeas/core/admin/space/SpaceInst.java (hunk 1 above)
Comment (SilverYoCha) [reply core-library/src/main/java/org/silverpeas/core/admin/space/SpaceInst.java:333]: Ho yes, now that it is only a list of identifier about components, it should be managed like you said into your first comment.
Head on the handlebars :-D

Comment (mmoqui) [reply core-library/src/main/java/org/silverpeas/core/admin/space/SpaceInst.java:333]: :-)

**Review by mmoqui — APPROVED** [review]

**Pull request merged**